acme/challenge.go

```diff
@@ -446,158 +441,178 @@ func wireOIDC01Validate(ctx context.Context, ch *Challenge, db DB, jwk *jose.JSO
 	}
 
-	err = stdin.Close()
-	if err != nil {
-		return WrapErrorISE(err, "error closing stdin")
+	order := orders[len(orders)-1]
+	if err := db.CreateDpopToken(ctx, order, map[string]any(*dpop)); err != nil {
+		return WrapErrorISE(err, "failed storing DPoP token")
 	}
 
-	err = cmd.Wait()
+	return nil
+}
+
+type cnf struct {
+	Kid string `json:"kid,omitempty"`
+}
+
+type wireAccessToken struct {
+	jose.Claims
+	Challenge  string `json:"chal,omitempty"`
+	Cnf        *cnf   `json:"cnf,omitempty"`
+	Proof      string `json:"proof,omitempty"`
+	ClientID   string `json:"client_id,omitempty"`
+	APIVersion int    `json:"api_version,omitempty"`
+	Scope      string `json:"scope,omitempty"`
+}
+
+type wireDpopToken map[string]any
+
+type verifyParams struct {
+	token      string
+	key        string
+	issuer     string
+	accountJWK *jose.JSONWebKey
+	wireID     wire.ID
+	challenge  *Challenge
+	t          time.Time
+}
+
+func parseAndVerifyWireAccessToken(v verifyParams) (*wireAccessToken, *wireDpopToken, error) {
+	k, err := pemutil.Parse([]byte(v.key)) // TODO(hs): move this to earlier in the configuration process? Do it once?
 	if err != nil {
-		return storeError(ctx, db, ch, true, NewError(ErrorRejectedIdentifierType, "error finishing validation: %s", err))
+		return nil, nil, fmt.Errorf("failed parsing public key: %w", err)
 	}
 
-	// Update and store the challenge.
-	ch.Status = StatusValid
-	ch.Error = nil
-	ch.ValidatedAt = clock.Now().Format(time.RFC3339)
-
-	if err = db.UpdateChallenge(ctx, ch); err != nil {
-		return WrapErrorISE(err, "error updating challenge")
+	pk, ok := k.(ed25519.PublicKey) // TODO(hs): allow more key types
+	if !ok {
+		return nil, nil, fmt.Errorf("unexpected type: %T", k)
 	}
 
-	parsedAccessToken, err := jose.ParseSigned(wireChallengePayload.AccessToken)
+	jwt, err := jose.ParseSigned(v.token)
 	if err != nil {
-		return WrapErrorISE(err, "invalid access token")
+		return nil, nil, fmt.Errorf("failed parsing token: %w", err)
 	}
-	access := make(map[string]interface{})
-	if err := parsedAccessToken.UnsafeClaimsWithoutVerification(&access); err != nil {
-		return WrapErrorISE(err, "failed parsing access token")
+
+	var accessToken wireAccessToken
+	if err = jwt.Claims(pk, &accessToken); err != nil {
+		return nil, nil, fmt.Errorf("failed validating Wire DPoP token claims: %w", err)
 	}
 
-	rawDpop, ok := access["proof"].(string)
-	if !ok {
-		return WrapErrorISE(err, "invalid dpop proof format in access token")
+	if err := accessToken.ValidateWithLeeway(jose.Expected{
+		Time:   v.t,
+		Issuer: v.issuer,
+	}, 360*time.Second); err != nil {
+		return nil, nil, fmt.Errorf("failed validation: %w", err)
 	}
 
-	parsedDpopToken, err := jose.ParseSigned(rawDpop)
+	rawKid, err := v.accountJWK.Thumbprint(crypto.SHA256)
 	if err != nil {
-		return WrapErrorISE(err, "invalid DPoP token")
+		return nil, nil, fmt.Errorf("failed to compute JWK thumbprint")
+	}
```

Comment:
Although it's probably the same, if the account has a kid `accountJWK.KeyID` we should use this one.